tests/mpi/triangulation_periodicity/test_1/test_1.cc

```diff
@@ -0,0 +1,324 @@
+ *       |           |           |           |           |
+ *       |           |           |           |           |
+ *       -------------------------------------------------
+ *       |           |     |     |     |     |           |
+ *       |           |     |     |     |     |           |
+ *       |           |     |     |     |     |           |
+ *       |           -------------------------           |
+ *       |           |     |     |     |     |           |
+ *       |           |     |     |     |     |           |
+ *       |           |     |     |     |     |           |
+ *       -------------------------------------------------
+ *                    Bottom periodic boundary
+ *
+ *
+ * =====================
+ * [ Passing criteria: ]
+ * =====================
+ *
+ * The test will be passed if the main three assertions are passed.
+ */
+
+using namespace dealii;
+
+template <int dim>
+struct periodicity_tests
+{
+  typedef TriaIterator<CellAccessor<dim>> cell_iterator;
+  periodicity_tests();
+  ~periodicity_tests();
+
+  unsigned refn_cycle;
+  MPI_Comm mpi_comm;
+  int comm_rank, comm_size;
+  parallel::distributed::Triangulation<dim> the_grid;
+
+  void refine_grid(const int &);
+  void write_grid();
+  void check_periodicity();
+};
+
+template <int dim>
+periodicity_tests<dim>::periodicity_tests()
+  : refn_cycle(0), mpi_comm(MPI_COMM_WORLD), the_grid(mpi_comm)
+{
+  assert(dim == 2 || dim == 3);
+  comm_rank = Utilities::MPI::this_mpi_process(mpi_comm);
+  comm_size = Utilities::MPI::n_mpi_processes(mpi_comm);
+  std::vector<unsigned> repeats(dim, 2);
+  Point<dim> p1, p2, periodic_transfer;
+  if (dim == 2)
+  {
+    p2 = { 16., 16. };
+    periodic_transfer = { 0.0, 16. };
+  }
+  if (dim == 3)
+  {
+    p2 = { 16., 16., 16. };
+    periodic_transfer = { 0.0, 16., 0.0 };
+  }
+  GridGenerator::subdivided_hyper_rectangle(the_grid, repeats, p1, p2, true);
+  std::vector<GridTools::PeriodicFacePair<cell_iterator>> periodic_faces;
+  GridTools::collect_periodic_faces(the_grid, 2, 3, 0, periodic_faces, periodic_transfer);
+  the_grid.add_periodicity(periodic_faces);
+}
+
+template <int dim>
+void periodicity_tests<dim>::refine_grid(const int &n = 1)
+{
+  if (n != 0 && refn_cycle == 0)
+  {
+    the_grid.refine_global(n);
+    refn_cycle += n;
+  }
+  else if (n != 0)
+  {
+    Point<dim> refn_point;
+    if (dim == 2)
+      refn_point = { 6.5, 15.5 };
+    if (dim == 3)
+      refn_point = { 6.5, 15.5, 6.5 };
+    for (unsigned i_refn = 0; i_refn < n; ++i_refn)
+    {
+      for (cell_iterator &&cell_it : the_grid.active_cell_iterators())
+      {
+        if (cell_it->is_locally_owned() && cell_it->point_inside(refn_point))
+        {
+          cell_it->set_refine_flag();
+          break;
+        }
+      }
+      the_grid.execute_coarsening_and_refinement();
+      ++refn_cycle;
+    }
+  }
+}
+
+template <int dim>
+void periodicity_tests<dim>::write_grid()
+{
```

Comment:
This is not really part of the test, but just for verifying the output manually, isn't it?